File: ai_core/utils/tracing.py
import uuid
import logging
from typing import Optional, Dict

logger = logging.getLogger(__name__)

class TraceContext:
    def __init__(self, trace_id: Optional[str] = None, span_id: Optional[str] = None):
        self.trace_id = trace_id or uuid.uuid4().hex
        self.span_id = span_id or uuid.uuid4().hex[:16]

    @classmethod
    def from_headers(cls, headers: Dict) -> 'TraceContext':
        """Extract trace context from request headers (W3C traceparent format)"""
        trace_id = headers.get('x-trace-id') or headers.get('X-Trace-Id')
        span_id = headers.get('x-span-id') or headers.get('X-Span-Id')

        if not trace_id:
            # Parse W3C traceparent header: 00-traceid-spanid-flags
            traceparent = headers.get('traceparent') or headers.get('Traceparent')
            if traceparent:
                parts = traceparent.split('-')
                if len(parts) >= 3:
                    trace_id = parts[1]
                    span_id = parts[2]

        return cls(trace_id, span_id)

    def to_headers(self) -> Dict[str, str]:
        """Convert trace context to headers for propagation"""
        return {
            'x-trace-id': self.trace_id,
            'x-span-id': self.span_id,
            'traceparent': f'00-{self.trace_id}-{self.span_id}-01',
        }


def init_tracing():
    """Initialize distributed tracing (W3C traceparent + custom headers)"""
    logger.info("Distributed tracing initialized (W3C traceparent + custom headers)")

File: backend/src/utils/tracing.js
const { v4: uuidv4 } = require('uuid');
const logger = require('../logger');

// Distributed tracing context: stores trace info for request correlation
class TraceContext {
  constructor(traceId, spanId, parentSpanId = null) {
    this.traceId = traceId;
    this.spanId = spanId;
    this.parentSpanId = parentSpanId;
    this.startTime = Date.now();
  }

  toHeaders() {
    return {
      'traceparent': `00-${this.traceId}-${this.spanId}-01`,
      'tracestate': '',
      'x-trace-id': this.traceId,
      'x-span-id': this.spanId,
    };
  }

  static fromHeaders(headers) {
    const traceId = headers['x-trace-id'] || headers['traceparent']?.split('-')[1] || uuidv4().replace(/-/g, '');
    const spanId = headers['x-span-id'] || headers['traceparent']?.split('-')[2] || uuidv4().replace(/-/g, '').slice(0, 16);
    return new TraceContext(traceId, spanId);
  }

  duration() {
    return Date.now() - this.startTime;
  }
}

// Middleware for Express: attach trace context to request
function traceMiddleware(req, res, next) {
  const trace = TraceContext.fromHeaders(req.headers);
  req.trace = trace;
  res.setHeader('X-Trace-Id', trace.traceId);
  res.setHeader('X-Span-Id', trace.spanId);
  next();
}

// Wrapper for axios to propagate trace context
function traceAxiosInterceptor(axiosInstance) {
  axiosInstance.interceptors.request.use((config) => {
    if (global.currentTrace) {
      const headers = global.currentTrace.toHeaders();
      Object.assign(config.headers || {}, headers);
    }
    return config;
  });

  axiosInstance.interceptors.response.use(
    (response) => response,
    (error) => Promise.reject(error),
  );
}

// Set current trace in async context (nodejs AsyncLocalStorage would be better but keeping it simple)
function setCurrentTrace(trace) {
  global.currentTrace = trace;
}

function getCurrentTrace() {
  return global.currentTrace;
}

function initTracing() {
  logger.info('Distributed tracing initialized (W3C traceparent + custom headers)');
}

module.exports = {
  TraceContext,
  traceMiddleware,
  traceAxiosInterceptor,
  setCurrentTrace,
  getCurrentTrace,
  initTracing,
};
